% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Untitled},
  pdfauthor={Daniel Redondo Sánchez},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering

\title{Untitled}
\author{Daniel Redondo Sánchez}
\date{9/5/2020}

\begin{document}
\maketitle

\hypertarget{descompresiuxf3n-de-ficheros}{%
\section{1. Descompresión de
ficheros}\label{descompresiuxf3n-de-ficheros}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# ----- Carga de paquetes -----}

\KeywordTok{library}\NormalTok{(BiocManager) }\CommentTok{# Para instalar KnowSeq}
\KeywordTok{library}\NormalTok{(KnowSeq)     }\CommentTok{# Para trabajar con genes - instalado a partir de .tar.gz de SWAD}
\KeywordTok{library}\NormalTok{(tictoc)      }\CommentTok{# Para medir tiempos con tic() y toc() a lo MATLAB }
\KeywordTok{library}\NormalTok{(R.utils)     }\CommentTok{# Para gunzip (descompresión de .gz)}
\KeywordTok{library}\NormalTok{(dplyr)       }\CommentTok{# Para select, filter, pipes, ...}
\KeywordTok{library}\NormalTok{(beepr)       }\CommentTok{# Para avisar con beeps cuando acaba un proceso}
\KeywordTok{library}\NormalTok{(caret)       }\CommentTok{# Para ml}
\KeywordTok{library}\NormalTok{(e1071)       }\CommentTok{# Para svm}
\KeywordTok{library}\NormalTok{(reshape)     }\CommentTok{# Para melt}
\KeywordTok{library}\NormalTok{(gplots)      }\CommentTok{# Para heatmaps}

\CommentTok{# ----- Preprocesamiento para adecuar ficheros a KnowSeq -----}

\CommentTok{# Lectura de sample_sheet}
\NormalTok{samplesInfo <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(}\StringTok{"gdc_sample_sheet.2020-05-09.tsv"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{"}\CharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }\DataTypeTok{header =}\NormalTok{ T)}

\CommentTok{# Se elimina ".counts.gz" de samplesInfo, porque countsToMatrix luego añade la extensión}
\NormalTok{samplesInfo[,}\DecValTok{2}\NormalTok{] <-}\StringTok{ }\KeywordTok{as.character}\NormalTok{(samplesInfo[,}\DecValTok{2}\NormalTok{])}
\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\KeywordTok{nrow}\NormalTok{(samplesInfo))\{}
\NormalTok{  samplesInfo[i,}\DecValTok{2}\NormalTok{] <-}\StringTok{ }\KeywordTok{substr}\NormalTok{(samplesInfo[i,}\DecValTok{2}\NormalTok{], }\DecValTok{1}\NormalTok{, }\KeywordTok{nchar}\NormalTok{(samplesInfo[i, }\DecValTok{2}\NormalTok{]) }\OperatorTok{-}\StringTok{ }\DecValTok{10}\NormalTok{)}
\NormalTok{\}}

\CommentTok{# Comprobación }
\KeywordTok{head}\NormalTok{(samplesInfo)}

\CommentTok{# Definición de parámetros}
\NormalTok{Run <-}\StringTok{ }\NormalTok{samplesInfo}\OperatorTok{$}\NormalTok{File.Name}
\NormalTok{Path <-}\StringTok{ }\NormalTok{samplesInfo}\OperatorTok{$}\NormalTok{File.ID}
\NormalTok{Class <-}\StringTok{ }\NormalTok{samplesInfo}\OperatorTok{$}\NormalTok{Sample.Type}

\KeywordTok{table}\NormalTok{(Class)}

\CommentTok{# Los casos metastásicos se recodifican a "Primary Tumor", aunque no sea lo ideal...}
\NormalTok{Class <-}\StringTok{ }\KeywordTok{gsub}\NormalTok{(}\StringTok{"Metastatic"}\NormalTok{, }\StringTok{"Primary Tumor"}\NormalTok{, Class)}
\KeywordTok{table}\NormalTok{(Class)}

\CommentTok{# Creación de dataframe}
\NormalTok{SamplesDataFrame <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(Run, Path, Class)}

\CommentTok{# Exportación a CSV de SamplesDataFrame}
\KeywordTok{setwd}\NormalTok{(}\DataTypeTok{dir =} \StringTok{"gdc_download_20200509_173000.089392//"}\NormalTok{)}
\KeywordTok{write.csv}\NormalTok{(SamplesDataFrame, }\DataTypeTok{file =} \StringTok{"SamplesDataFrame.csv"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{preprocesamiento}{%
\section{2. Preprocesamiento}\label{preprocesamiento}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# ----- Ruta de trabajo -----}

\CommentTok{# Windows}
\KeywordTok{setwd}\NormalTok{(}\StringTok{"C:/Users/dredondo/Dropbox/Transporte_interno/Máster/Ciencia de Datos/TFM/Análisis preliminar/descarga_20200509"}\NormalTok{)}
\CommentTok{# Mac}
\CommentTok{#setwd("/Users/daniel/Dropbox/Transporte_interno/Máster/Ciencia de Datos/TFM/Análisis preliminar/descarga_2020050")}

\CommentTok{# ----- Carga de paquetes -----}

\KeywordTok{library}\NormalTok{(BiocManager) }\CommentTok{# Para instalar KnowSeq}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Bioconductor version 3.11 (BiocManager 1.30.10), ?BiocManager::install for help
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(KnowSeq)     }\CommentTok{# Para trabajar con genes - instalado a partir de .tar.gz de SWAD}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Loading required package: cqn
\end{verbatim}

\begin{verbatim}
## Loading required package: mclust
\end{verbatim}

\begin{verbatim}
## Package 'mclust' version 5.4.6
## Type 'citation("mclust")' for citing this R package in publications.
\end{verbatim}

\begin{verbatim}
## Loading required package: nor1mix
\end{verbatim}

\begin{verbatim}
## Loading required package: preprocessCore
\end{verbatim}

\begin{verbatim}
## Loading required package: splines
\end{verbatim}

\begin{verbatim}
## Loading required package: quantreg
\end{verbatim}

\begin{verbatim}
## Loading required package: SparseM
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'SparseM'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:base':
## 
##     backsolve
\end{verbatim}

\begin{verbatim}
## ##############################################################################
## Pathview is an open source software package distributed under GNU General
## Public License version 3 (GPLv3). Details of GPLv3 is available at
## http://www.gnu.org/licenses/gpl-3.0.html. Particullary, users are required to
## formally cite the original Pathview paper (not just mention it) in publications
## or products. For details, do citation("pathview") within R.
## 
## The pathview downloads and uses KEGG data. Non-academic uses may require a KEGG
## license agreement (details at http://www.kegg.jp/kegg/legal.html).
## ##############################################################################
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(tictoc)      }\CommentTok{# Para medir tiempos con tic() y toc() a lo MATLAB }
\KeywordTok{library}\NormalTok{(R.utils)     }\CommentTok{# Para gunzip (descompresión de .gz)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Loading required package: R.oo
\end{verbatim}

\begin{verbatim}
## Loading required package: R.methodsS3
\end{verbatim}

\begin{verbatim}
## R.methodsS3 v1.8.0 (2020-02-14 07:10:20 UTC) successfully loaded. See ?R.methodsS3 for help.
\end{verbatim}

\begin{verbatim}
## R.oo v1.23.0 successfully loaded. See ?R.oo for help.
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'R.oo'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:R.methodsS3':
## 
##     throw
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:methods':
## 
##     getClasses, getMethods
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:base':
## 
##     attach, detach, load, save
\end{verbatim}

\begin{verbatim}
## R.utils v2.9.2 successfully loaded. See ?R.utils for help.
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'R.utils'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:utils':
## 
##     timestamp
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:base':
## 
##     cat, commandArgs, getOption, inherits, isOpen, nullfile, parse,
##     warnings
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(dplyr)       }\CommentTok{# Para select, filter, pipes, ...}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Attaching package: 'dplyr'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:stats':
## 
##     filter, lag
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(beepr)       }\CommentTok{# Para avisar con beeps cuando acaba un proceso}
\KeywordTok{library}\NormalTok{(caret)       }\CommentTok{# Para ml}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Loading required package: lattice
\end{verbatim}

\begin{verbatim}
## Loading required package: ggplot2
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(e1071)       }\CommentTok{# Para svm}
\KeywordTok{library}\NormalTok{(reshape)     }\CommentTok{# Para melt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Attaching package: 'reshape'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:dplyr':
## 
##     rename
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(gplots)      }\CommentTok{# Para heatmaps}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Attaching package: 'gplots'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:stats':
## 
##     lowess
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# ----- Preprocesamiento para adecuar ficheros a KnowSeq -----}

\CommentTok{# Lectura de sample_sheet}
\NormalTok{samplesInfo <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(}\StringTok{"gdc_sample_sheet.2020-05-09.tsv"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{"}\CharTok{\textbackslash{}t}\StringTok{"}\NormalTok{, }\DataTypeTok{header =}\NormalTok{ T)}

\CommentTok{# Se elimina ".counts.gz" de samplesInfo, porque countsToMatrix luego añade la extensión}
\NormalTok{samplesInfo[,}\DecValTok{2}\NormalTok{] <-}\StringTok{ }\KeywordTok{as.character}\NormalTok{(samplesInfo[,}\DecValTok{2}\NormalTok{])}
\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\KeywordTok{nrow}\NormalTok{(samplesInfo))\{}
\NormalTok{  samplesInfo[i,}\DecValTok{2}\NormalTok{] <-}\StringTok{ }\KeywordTok{substr}\NormalTok{(samplesInfo[i,}\DecValTok{2}\NormalTok{], }\DecValTok{1}\NormalTok{, }\KeywordTok{nchar}\NormalTok{(samplesInfo[i, }\DecValTok{2}\NormalTok{]) }\OperatorTok{-}\StringTok{ }\DecValTok{10}\NormalTok{)}
\NormalTok{\}}

\CommentTok{# Comprobación }
\KeywordTok{head}\NormalTok{(samplesInfo)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##                                File.ID
## 1 6da17f3f-d4ee-4ace-a6f7-702454c8f0e9
## 2 fdf73b53-a45b-4f06-8418-19896fc3d076
## 3 7e374f79-6f9b-4034-b4cb-d71b7404682a
## 4 30a0d8dd-ba66-41d0-9a32-24de4e1a28c7
## 5 7fa837c1-9fb9-4c4d-ac22-4bd3a058acf3
## 6 4d4602d2-35bf-492b-8c6a-b3b840ca2270
##                                    File.Name           Data.Category
## 1 b9ab7393-4abb-41ec-9d55-a3dc846c4a93.htseq Transcriptome Profiling
## 2 aec2e0c7-4792-41af-873c-3f3a53ec6d38.htseq Transcriptome Profiling
## 3 4172e3f8-3578-4f33-9168-6f8c2b8d0783.htseq Transcriptome Profiling
## 4 cdaedb53-612c-4e7a-8ffe-348944b94e0c.htseq Transcriptome Profiling
## 5 5aed2227-1f31-4159-9eed-430bc45c61dc.htseq Transcriptome Profiling
## 6 a2a33be8-232b-44bf-a003-349017a5bc5a.htseq Transcriptome Profiling
##                        Data.Type Project.ID      Case.ID        Sample.ID
## 1 Gene Expression Quantification  TCGA-PAAD TCGA-HZ-7926 TCGA-HZ-7926-01A
## 2 Gene Expression Quantification  TCGA-PAAD TCGA-HZ-8001 TCGA-HZ-8001-01A
## 3 Gene Expression Quantification  TCGA-PAAD TCGA-3E-AAAZ TCGA-3E-AAAZ-01A
## 4 Gene Expression Quantification  TCGA-PAAD TCGA-LB-A7SX TCGA-LB-A7SX-01A
## 5 Gene Expression Quantification  TCGA-PAAD TCGA-2L-AAQM TCGA-2L-AAQM-01A
## 6 Gene Expression Quantification  TCGA-PAAD TCGA-RB-A7B8 TCGA-RB-A7B8-01A
##     Sample.Type
## 1 Primary Tumor
## 2 Primary Tumor
## 3 Primary Tumor
## 4 Primary Tumor
## 5 Primary Tumor
## 6 Primary Tumor
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Definición de parámetros}
\NormalTok{Run <-}\StringTok{ }\NormalTok{samplesInfo}\OperatorTok{$}\NormalTok{File.Name}
\NormalTok{Path <-}\StringTok{ }\NormalTok{samplesInfo}\OperatorTok{$}\NormalTok{File.ID}
\NormalTok{Class <-}\StringTok{ }\NormalTok{samplesInfo}\OperatorTok{$}\NormalTok{Sample.Type}

\KeywordTok{table}\NormalTok{(Class)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Class
##          Metastatic       Primary Tumor Solid Tissue Normal 
##                   1                 177                   4
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Los casos metastásicos se recodifican a "Primary Tumor", aunque no sea lo ideal...}
\NormalTok{Class <-}\StringTok{ }\KeywordTok{gsub}\NormalTok{(}\StringTok{"Metastatic"}\NormalTok{, }\StringTok{"Primary Tumor"}\NormalTok{, Class)}
\KeywordTok{table}\NormalTok{(Class)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Class
##       Primary Tumor Solid Tissue Normal 
##                 178                   4
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Creación de dataframe}
\NormalTok{SamplesDataFrame <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(Run, Path, Class)}

\CommentTok{# Exportación a CSV de SamplesDataFrame}
\KeywordTok{setwd}\NormalTok{(}\DataTypeTok{dir =} \StringTok{"gdc_download_20200509_173000.089392//"}\NormalTok{)}
\KeywordTok{write.csv}\NormalTok{(SamplesDataFrame, }\DataTypeTok{file =} \StringTok{"SamplesDataFrame.csv"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{anuxe1lisis}{%
\section{3. Análisis}\label{anuxe1lisis}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# ----- Carga de paquetes -----}

\KeywordTok{library}\NormalTok{(BiocManager) }\CommentTok{# Para instalar KnowSeq}
\KeywordTok{library}\NormalTok{(KnowSeq)     }\CommentTok{# Para trabajar con genes - instalado a partir de .tar.gz de SWAD}
\KeywordTok{library}\NormalTok{(tictoc)      }\CommentTok{# Para medir tiempos con tic() y toc() a lo MATLAB }
\KeywordTok{library}\NormalTok{(R.utils)     }\CommentTok{# Para gunzip (descompresión de .gz)}
\KeywordTok{library}\NormalTok{(dplyr)       }\CommentTok{# Para select, filter, pipes, ...}
\KeywordTok{library}\NormalTok{(beepr)       }\CommentTok{# Para avisar con beeps cuando acaba un proceso}
\KeywordTok{library}\NormalTok{(caret)       }\CommentTok{# Para ml}
\KeywordTok{library}\NormalTok{(e1071)       }\CommentTok{# Para svm}
\KeywordTok{library}\NormalTok{(reshape)     }\CommentTok{# Para melt}
\KeywordTok{library}\NormalTok{(gplots)      }\CommentTok{# Para heatmaps}

\CommentTok{# ----- Unificación de ficheros en formato matriz -----}

\KeywordTok{tic}\NormalTok{(}\StringTok{"countsToMatrix"}\NormalTok{)}
\NormalTok{countsInformation <-}\StringTok{ }\KeywordTok{countsToMatrix}\NormalTok{(}\StringTok{"SamplesDataFrame.csv"}\NormalTok{, }\DataTypeTok{extension =} \StringTok{"counts"}\NormalTok{)}
\KeywordTok{toc}\NormalTok{()}

\CommentTok{# ----- La matriz de cuentas y las etiquetas se guardan -----}

\NormalTok{countsMatrix <-}\StringTok{ }\NormalTok{countsInformation}\OperatorTok{$}\NormalTok{countsMatrix}
\NormalTok{labels <-}\StringTok{ }\NormalTok{countsInformation}\OperatorTok{$}\NormalTok{labels}

\CommentTok{# ----- Se descargan los nombres de los genes ----- }

\NormalTok{myAnnotation <-}\StringTok{ }\KeywordTok{getGenesAnnotation}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(countsMatrix))}

\CommentTok{# ----- Cálculo de matriz de expresión de genes -----}

\KeywordTok{tic}\NormalTok{(}\StringTok{"calculateGeneExpressionValues"}\NormalTok{)}
\NormalTok{expressionMatrix <-}\StringTok{ }\KeywordTok{calculateGeneExpressionValues}\NormalTok{(countsMatrix, myAnnotation, }\DataTypeTok{genesNames =} \OtherTok{TRUE}\NormalTok{)}
\KeywordTok{toc}\NormalTok{()}

\CommentTok{# ----- Boxplots de la expresión para todos los genes -----}

\KeywordTok{dataPlot}\NormalTok{(expressionMatrix, labels, }\DataTypeTok{mode =} \StringTok{"boxplot"}\NormalTok{, }\DataTypeTok{colours =} \KeywordTok{c}\NormalTok{(}\StringTok{"blue"}\NormalTok{, }\StringTok{"red"}\NormalTok{), }\DataTypeTok{toPNG =} \OtherTok{TRUE}\NormalTok{)}
\KeywordTok{dataPlot}\NormalTok{(expressionMatrix, labels, }\DataTypeTok{mode =} \StringTok{"orderedBoxplot"}\NormalTok{, }\DataTypeTok{colours =} \KeywordTok{c}\NormalTok{(}\StringTok{"blue"}\NormalTok{, }\StringTok{"red"}\NormalTok{), }\DataTypeTok{toPNG =} \OtherTok{TRUE}\NormalTok{)}

\CommentTok{# ----- Controlando por el efecto batch -----}

\KeywordTok{tic}\NormalTok{(}\StringTok{"batchEffectRemoval"}\NormalTok{) }
\NormalTok{svaMod <-}\StringTok{ }\KeywordTok{batchEffectRemoval}\NormalTok{(expressionMatrix, }\KeywordTok{as.factor}\NormalTok{(labels), }\DataTypeTok{method =} \StringTok{"sva"}\NormalTok{)}
\KeywordTok{toc}\NormalTok{()}

\CommentTok{# ----- Extracción de DEG (Expresión Diferencial de Genes) -----}

\KeywordTok{tic}\NormalTok{(}\StringTok{"DEGsExtraction"}\NormalTok{) }
\NormalTok{DEGsInformation <-}\StringTok{ }\KeywordTok{DEGsExtraction}\NormalTok{(expressionMatrix, }\KeywordTok{as.factor}\NormalTok{(labels),}
                                  \CommentTok{# Default parameters}
                                  \DataTypeTok{lfc =} \FloatTok{1.0}\NormalTok{, }\DataTypeTok{pvalue =} \FloatTok{0.1}\NormalTok{,}
                                  \CommentTok{# Ajuste por batchEffect}
                                  \DataTypeTok{svaCorrection =} \OtherTok{TRUE}\NormalTok{, }\DataTypeTok{svaMod =}\NormalTok{ svaMod)}
\KeywordTok{toc}\NormalTok{()}

\CommentTok{# Número de genes extraídos}
\KeywordTok{print}\NormalTok{(}\KeywordTok{nrow}\NormalTok{(DEGsInformation}\OperatorTok{$}\NormalTok{DEGsMatrix))}

\NormalTok{topTable <-}\StringTok{ }\NormalTok{DEGsInformation}\OperatorTok{$}\NormalTok{Table}
\NormalTok{DEGsMatrix <-}\StringTok{ }\NormalTok{DEGsInformation}\OperatorTok{$}\NormalTok{DEGsMatrix}

\KeywordTok{head}\NormalTok{(topTable) }\CommentTok{# Si la primera columna es negativa, infrarepresentado}
\CommentTok{# positiva = gen sobreexpresado}



\CommentTok{# Se traspone la matriz}
\NormalTok{DEGsMatrixML <-}\StringTok{ }\KeywordTok{t}\NormalTok{(DEGsMatrix)}

\CommentTok{# ----- Representación de DEG -----}

\CommentTok{# Boxplots para todas las muestras de los primeros genes}
\KeywordTok{dataPlot}\NormalTok{(DEGsMatrix[}\DecValTok{1}\OperatorTok{:}\DecValTok{11}\NormalTok{, ], labels, }\DataTypeTok{mode =} \StringTok{"genesBoxplot"}\NormalTok{, }\DataTypeTok{toPNG =} \OtherTok{TRUE}\NormalTok{)}

\CommentTok{# Mapa de calor para todas las muestras de los primeros genes}
\KeywordTok{dataPlot}\NormalTok{(DEGsMatrix[}\DecValTok{1}\OperatorTok{:}\DecValTok{11}\NormalTok{, ], labels, }\DataTypeTok{mode =} \StringTok{"heatmap"}\NormalTok{, }\DataTypeTok{toPNG =} \OtherTok{TRUE}\NormalTok{)}

\CommentTok{# ----- Partición entrenamiento-test -----}

\CommentTok{# Partición 50% / 50% con balanceo de clase}
\NormalTok{indices <-}\StringTok{ }\KeywordTok{createDataPartition}\NormalTok{(SamplesDataFrame}\OperatorTok{$}\NormalTok{Class, }\DataTypeTok{p =} \FloatTok{.5}\NormalTok{, }\DataTypeTok{list =} \OtherTok{FALSE}\NormalTok{)}
\NormalTok{particion <-}\StringTok{ }\KeywordTok{list}\NormalTok{(}\DataTypeTok{training =}\NormalTok{ DEGsMatrixML[indices, ], }\DataTypeTok{test =}\NormalTok{ DEGsMatrixML[}\OperatorTok{-}\NormalTok{indices, ])}

\CommentTok{# Conjuntos}
\NormalTok{particion.entrenamiento <-}\StringTok{ }\NormalTok{particion}\OperatorTok{$}\NormalTok{training}
\NormalTok{particion.test <-}\StringTok{ }\NormalTok{particion}\OperatorTok{$}\NormalTok{test}

\CommentTok{# Etiquetas}
\NormalTok{labels_train <-}\StringTok{ }\NormalTok{SamplesDataFrame}\OperatorTok{$}\NormalTok{Class[indices]}
\NormalTok{labels_test  <-}\StringTok{ }\NormalTok{SamplesDataFrame}\OperatorTok{$}\NormalTok{Class[}\OperatorTok{-}\NormalTok{indices]}

\CommentTok{# Número de casos}
\CommentTok{# Train}
\KeywordTok{table}\NormalTok{(labels_train)}
\CommentTok{# Test}
\KeywordTok{table}\NormalTok{(labels_test)}

\CommentTok{# Verificar balanceo de clase en entrenamiento y test}
\CommentTok{# Train}
\NormalTok{labels_train }\OperatorTok{%>%}\StringTok{ }\NormalTok{table }\OperatorTok{%>%}\StringTok{ }\NormalTok{prop.table }\OperatorTok{%>%}\StringTok{ }\KeywordTok{round}\NormalTok{(}\DecValTok{3}\NormalTok{) }\OperatorTok{*}\StringTok{ }\DecValTok{100}
\CommentTok{# Test}
\NormalTok{labels_test }\OperatorTok{%>%}\StringTok{ }\NormalTok{table }\OperatorTok{%>%}\StringTok{ }\NormalTok{prop.table }\OperatorTok{%>%}\StringTok{ }\KeywordTok{round}\NormalTok{(}\DecValTok{3}\NormalTok{) }\OperatorTok{*}\StringTok{ }\DecValTok{100}

\CommentTok{# ----- Selección de características -----}

\CommentTok{# Método mRMR (mínima redundancia, máxima relevancia)}
\NormalTok{mrmrRanking <-}\StringTok{ }\KeywordTok{featureSelection}\NormalTok{(particion.entrenamiento, labels_train, }\KeywordTok{colnames}\NormalTok{(particion.entrenamiento),}
                                \DataTypeTok{mode =} \StringTok{"mrmr"}\NormalTok{)}

\CommentTok{# Método random forest}
\NormalTok{rfRanking <-}\StringTok{ }\KeywordTok{featureSelection}\NormalTok{(particion.entrenamiento, labels_train, }\KeywordTok{colnames}\NormalTok{(particion.entrenamiento),}
                              \DataTypeTok{mode =} \StringTok{"rf"}\NormalTok{)}

\CommentTok{# Método Disease association ranking (en base a scores obtenidos en la literatura)}
\NormalTok{daRanking <-}\StringTok{ }\KeywordTok{featureSelection}\NormalTok{(particion.entrenamiento, labels_train, }\KeywordTok{colnames}\NormalTok{(particion.entrenamiento),}
                              \DataTypeTok{mode =} \StringTok{"da"}\NormalTok{, }\DataTypeTok{disease =} \StringTok{"pancreatic cancer"}\NormalTok{)}

\CommentTok{# -----  Sobreescribir la función dataPlot con la nueva función que pinta líneas discontinuas -----}

\KeywordTok{source}\NormalTok{(}\StringTok{"../../Funciones_actualizadas_KnowSeq/dataPlot.R"}\NormalTok{)}

\CommentTok{# ----- Resultados de SVM con validación cruzada para cada método de selección de características -----}

\NormalTok{numero_folds <-}\StringTok{ }\DecValTok{2}

\CommentTok{# mRMR}
\NormalTok{results_cv_svm_mrmr <-}\StringTok{ }\KeywordTok{svm_CV}\NormalTok{(particion.entrenamiento, labels_train, }\KeywordTok{names}\NormalTok{(mrmrRanking),}
                              \DataTypeTok{numFold =}\NormalTok{ numero_folds)}
\NormalTok{results_cv_svm_mrmr}\OperatorTok{$}\NormalTok{bestParameters}

\CommentTok{# random forest}
\NormalTok{results_cv_svm_rf <-}\StringTok{ }\KeywordTok{svm_CV}\NormalTok{(particion.entrenamiento, labels_train, rfRanking,}
                            \DataTypeTok{numFold =}\NormalTok{ numero_folds)}
\NormalTok{results_cv_svm_rf}\OperatorTok{$}\NormalTok{bestParameters}

\CommentTok{# disease association}
\NormalTok{results_cv_svm_da <-}\StringTok{ }\KeywordTok{svm_CV}\NormalTok{(particion.entrenamiento, labels_train, }\KeywordTok{names}\NormalTok{(daRanking),}
                            \DataTypeTok{numFold =}\NormalTok{ numero_folds)}
\NormalTok{results_cv_svm_da}\OperatorTok{$}\NormalTok{bestParameters}

\CommentTok{# ----- Resultados de SVM gráficamente -----}

\CommentTok{# Plotting the accuracy of all the folds evaluated in the CV process}
\CommentTok{#png(filename = "MRMR.png")}
\KeywordTok{dataPlot}\NormalTok{(results_cv_svm_mrmr}\OperatorTok{$}\NormalTok{accMatrix[, }\DecValTok{1}\OperatorTok{:}\DecValTok{11}\NormalTok{], }\DataTypeTok{mode =} \StringTok{"classResults"}\NormalTok{,}
         \DataTypeTok{main =} \StringTok{"mRMR - Accuracy for each fold"}\NormalTok{, }\DataTypeTok{xlab =} \StringTok{"Genes"}\NormalTok{, }\DataTypeTok{ylab =} \StringTok{"Accuracy"}\NormalTok{, }\DataTypeTok{toPNG =} \OtherTok{TRUE}\NormalTok{)}
\CommentTok{#dev.off()}

\CommentTok{# Plotting the accuracy of all the folds evaluated in the CV process}
\KeywordTok{dataPlot}\NormalTok{(results_cv_svm_rf}\OperatorTok{$}\NormalTok{accMatrix[, }\DecValTok{1}\OperatorTok{:}\DecValTok{9}\NormalTok{], }\DataTypeTok{colours =} \KeywordTok{rainbow}\NormalTok{(numero_folds), }\DataTypeTok{mode =} \StringTok{"classResults"}\NormalTok{,}
         \DataTypeTok{main =} \StringTok{"rf - Accuracy for each fold"}\NormalTok{, }\DataTypeTok{xlab =} \StringTok{"Genes"}\NormalTok{, }\DataTypeTok{ylab =} \StringTok{"Accuracy"}\NormalTok{, }\DataTypeTok{toPNG =} \OtherTok{TRUE}\NormalTok{)}

\CommentTok{# Plotting the accuracy of all the folds evaluated in the CV process}
\KeywordTok{dataPlot}\NormalTok{(results_cv_svm_da}\OperatorTok{$}\NormalTok{accMatrix[, }\DecValTok{1}\OperatorTok{:}\DecValTok{9}\NormalTok{], }\DataTypeTok{colours =} \KeywordTok{rainbow}\NormalTok{(numero_folds), }\DataTypeTok{mode =} \StringTok{"classResults"}\NormalTok{,}
         \DataTypeTok{main =} \StringTok{"da - Accuracy for each fold"}\NormalTok{, }\DataTypeTok{xlab =} \StringTok{"Genes"}\NormalTok{, }\DataTypeTok{ylab =} \StringTok{"Accuracy"}\NormalTok{, }\DataTypeTok{toPNG =} \OtherTok{TRUE}\NormalTok{)}

\CommentTok{# --- Mejor método en CV ---}
\NormalTok{genes_a_usar <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{2}\NormalTok{, }\DecValTok{3}\NormalTok{, }\DecValTok{5}\NormalTok{)}

\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\KeywordTok{length}\NormalTok{(genes_a_usar))\{}
\NormalTok{  precisiones <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\KeywordTok{mean}\NormalTok{(results_cv_svm_mrmr}\OperatorTok{$}\NormalTok{accMatrix[, genes_a_usar[i]]),}
                   \KeywordTok{mean}\NormalTok{(results_cv_svm_rf}\OperatorTok{$}\NormalTok{accMatrix[, genes_a_usar[i]]),}
                   \KeywordTok{mean}\NormalTok{(results_cv_svm_da}\OperatorTok{$}\NormalTok{accMatrix[, genes_a_usar[i]]))}
  
  \KeywordTok{names}\NormalTok{(precisiones) <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"MRMR"}\NormalTok{, }\StringTok{"RF"}\NormalTok{, }\StringTok{"DA"}\NormalTok{)}
  \KeywordTok{cat}\NormalTok{(}\KeywordTok{paste}\NormalTok{(}\StringTok{"Precisiones para"}\NormalTok{, genes_a_usar[i], }\StringTok{"gen(es):}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{))}
  \KeywordTok{print}\NormalTok{(precisiones)  }
  \KeywordTok{cat}\NormalTok{(}\StringTok{"-----------------------------}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\}}

\CommentTok{# ----- Resultados de SVM train-test -----}

\CommentTok{# mRMR}
\NormalTok{results_svm_mrmr <-}\StringTok{ }\KeywordTok{svm_test}\NormalTok{(}\DataTypeTok{train =}\NormalTok{ particion.entrenamiento, labels_train,}
                             \DataTypeTok{test =}\NormalTok{ particion.test, labels_test, }\KeywordTok{names}\NormalTok{(mrmrRanking),}
                             \DataTypeTok{bestParameters =}\NormalTok{ results_cv_svm_mrmr}\OperatorTok{$}\NormalTok{bestParameters)}

\CommentTok{# random forest}
\NormalTok{results_svm_rf <-}\StringTok{ }\KeywordTok{svm_test}\NormalTok{(}\DataTypeTok{train =}\NormalTok{ particion.entrenamiento, labels_train,}
                           \DataTypeTok{test =}\NormalTok{ particion.test, labels_test, rfRanking,}
                           \DataTypeTok{bestParameters =}\NormalTok{ results_cv_svm_rf}\OperatorTok{$}\NormalTok{bestParameters)}

\CommentTok{# disease association}
\NormalTok{results_svm_da <-}\StringTok{ }\KeywordTok{svm_test}\NormalTok{(}\DataTypeTok{train =}\NormalTok{ particion.entrenamiento, labels_train,}
                           \DataTypeTok{test =}\NormalTok{ particion.test, labels_test, }\KeywordTok{names}\NormalTok{(daRanking),}
                           \DataTypeTok{bestParameters =}\NormalTok{ results_cv_svm_da}\OperatorTok{$}\NormalTok{bestParameters)}

\CommentTok{# ----- Matrices de confusión -----}

\CommentTok{# Matriz de confusión}
\NormalTok{results_svm_mrmr}\OperatorTok{$}\NormalTok{cfMats[[}\DecValTok{1}\NormalTok{]]}\OperatorTok{$}\NormalTok{table}
\NormalTok{results_svm_rf}\OperatorTok{$}\NormalTok{cfMats[[}\DecValTok{1}\NormalTok{]]}\OperatorTok{$}\NormalTok{table}
\NormalTok{results_svm_da}\OperatorTok{$}\NormalTok{cfMats[[}\DecValTok{1}\NormalTok{]]}\OperatorTok{$}\NormalTok{table}

\CommentTok{# Gráficamente}
\KeywordTok{dataPlot}\NormalTok{(results_svm_mrmr}\OperatorTok{$}\NormalTok{cfMats[[}\DecValTok{1}\NormalTok{]]}\OperatorTok{$}\NormalTok{table, labels_test,}
         \DataTypeTok{mode =} \StringTok{"confusionMatrix"}\NormalTok{)}
\KeywordTok{dataPlot}\NormalTok{(results_svm_rf}\OperatorTok{$}\NormalTok{cfMats[[}\DecValTok{1}\NormalTok{]]}\OperatorTok{$}\NormalTok{table, labels_test,}
         \DataTypeTok{mode =} \StringTok{"confusionMatrix"}\NormalTok{)}
\KeywordTok{dataPlot}\NormalTok{(results_svm_da}\OperatorTok{$}\NormalTok{cfMats[[}\DecValTok{1}\NormalTok{]]}\OperatorTok{$}\NormalTok{table, labels_test,}
         \DataTypeTok{mode =} \StringTok{"confusionMatrix"}\NormalTok{)}

\CommentTok{# ----- Mejor método hallado en validación cruzada (rf, 5 genes) -----}

\KeywordTok{dataPlot}\NormalTok{(results_svm_rf}\OperatorTok{$}\NormalTok{cfMats[[}\DecValTok{5}\NormalTok{]]}\OperatorTok{$}\NormalTok{table, labels_test,}
         \DataTypeTok{mode =} \StringTok{"confusionMatrix"}\NormalTok{)}

\CommentTok{# ----- Descarga de información sobre enfermedades relacionadas con DEGs -----}

\CommentTok{# Se recuperan las 20 enfermedades más importantes vinculadas}
\CommentTok{# Se usan los 5 genes más relevantes usando rf, el mejor método encontrado anteriormente}
\NormalTok{diseases <-}\StringTok{ }\KeywordTok{DEGsToDiseases}\NormalTok{(rfRanking[}\DecValTok{1}\OperatorTok{:}\DecValTok{5}\NormalTok{], }\DataTypeTok{getEvidences =} \OtherTok{TRUE}\NormalTok{, }\DataTypeTok{size =} \DecValTok{20}\NormalTok{)}
\NormalTok{diseases}

\CommentTok{# Extracción de todas las enfermedades relacionadas con los 5 genes}
\NormalTok{enfermedades <-}\StringTok{ }\KeywordTok{c}\NormalTok{()}
\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{5}\NormalTok{)\{}
  \CommentTok{# Se extraen las enfermedades relacionadas}
\NormalTok{  enfermedades <-}\StringTok{ }\KeywordTok{c}\NormalTok{(enfermedades, diseases[[i]]}\OperatorTok{$}\NormalTok{summary[, }\DecValTok{1}\NormalTok{])}
  \CommentTok{# Se eliminan duplicados}
\NormalTok{  enfermedades <-}\StringTok{ }\KeywordTok{unique}\NormalTok{(enfermedades)}
\NormalTok{\}}
\NormalTok{enfermedades}

\CommentTok{# Buscar cáncer}
\NormalTok{enfermedades[}\KeywordTok{grep}\NormalTok{(}\DataTypeTok{pattern =} \StringTok{"cancer"}\NormalTok{, }\DataTypeTok{x =}\NormalTok{ enfermedades)]}
\NormalTok{enfermedades[}\KeywordTok{grep}\NormalTok{(}\DataTypeTok{pattern =} \StringTok{"neopl"}\NormalTok{, }\DataTypeTok{x =}\NormalTok{ enfermedades)]}
\NormalTok{enfermedades[}\KeywordTok{grep}\NormalTok{(}\DataTypeTok{pattern =} \StringTok{"panc"}\NormalTok{, }\DataTypeTok{x =}\NormalTok{ enfermedades)]}

\CommentTok{# ----- Ontología de genes -----}

\CommentTok{# labelsGo <- gsub("Solid Tissue Normal",0,labels)}
\CommentTok{# labelsGo <- gsub("Primary Tumor",1,labelsGo)}
\CommentTok{# GOsMatrix <- geneOntologyEnrichment(rownames(DEGsMatrix)[1], labelsGo)}
\CommentTok{# str(GOsMatrix, max.level = 10)}


\CommentTok{# ----- Pathway visualization (no encuentra ningún pathway, la extraigo de KEGG finalmente) -----}
\CommentTok{# Annotation_counts <- getAnnotationFromEnsembl(rownames(countsMatrix))}
\CommentTok{# Annotation_DEGsMatrix <- Annotation_counts %>% filter(external_gene_name %in% rownames(DEGsMatrix))}
\CommentTok{# }
\CommentTok{# expressionMatrix2 <- expressionMatrix}
\CommentTok{# Annotation_expressionMatrix <- Annotation_counts %>% filter(external_gene_name %in% rownames(expressionMatrix2))}
\CommentTok{# }
\CommentTok{# DEGsPathwayVisualization(DEGsMatrix = DEGsMatrix, DEGsAnnotation = Annotation_DEGsMatrix,}
\CommentTok{#                          expressionMatrix = expressionMatrix2, expressionAnnotation = Annotation_expressionMatrix,}
\CommentTok{#                          labels = labels)}
\CommentTok{# }
\end{Highlighting}
\end{Shaded}

\end{document}
